command_line/generate_mask.py

```diff
@@ -85,55 +86,70 @@
 def generate_mask(experiments, params):
     # type: (ExperimentList, phil.scope_extract) -> Tuple[Masks, Optional[ExperimentList]]
     """
-    Generate a pixel mask for each image in an experiment.
+    Generate a pixel mask for each imageset in an experiment list.
 
-    Use the masking parameters :param:`params` and an experiment in the experiment
-    list :param:`experiments` to define pixel masks for the associated imageset.
-    The masks are generated using :mod:`dials.util.masking`.
+    Use the masking parameters :param:`params` and the experiments in the experiment
+    list :param:`experiments` to define a pixel mask for each of the associated
+    imagesets.  The masks are generated using :mod:`dials.util.masking`.
 
     The masks will be saved to disk at the location specified by
-    :attr:`params.output.mask`.  Optionally, if a path
-    :attr:`params.output.experiments` is set, a modified copy of :param:`experiments`
-    with the masks applied will be saved to that location.
+    :attr:`params.output.mask`.  If the experiment list contains more than one
+    imageset, multiple mask files will be produced, with filenames differentiated by
+    an appended number.  Optionally, if a path :attr:`params.output.experiments` is
+    set, a modified copy of :param:`experiments` with the masks applied will be saved
+    to that location.
 
     Args:
         experiments: An experiment list containing only one imageset.
         params: Masking parameters, having the structure defined in
             :data:`phil_scope`.
 
     Returns:
-        A tuple containing the generated pixel masks.
+        A list of masks, one for each imageset.
 
         A copy of :param:`experiments` with the masks applied (optional,
         only returned if :attr:`params.output.experiments` is set).
     """
     imagesets = experiments.imagesets()
-    assert len(imagesets) == 1, "Only single imagesets supported"
-
-    imageset = imagesets[0]
+    masks = []
+
+    # Specify the formatting of the output filenames
+    num_imagesets = len(imagesets)
+    pad = len("{:d}".format(num_imagesets))
+    # If there is more than one imageset, append a number to each output mask filename
+    filenames = [
+        "_{:0{:d}d}".format(i + 1, pad).join(os.path.splitext(params.output.mask))
+        if num_imagesets > 1
+        else params.output.mask
+        for i in range(num_imagesets)
+    ]
```

Comment:
Had to do some mental contortions for this loop - both the `join(os.path.splitext(...` is a bit of an unusual way to do that (though, clever!). Also found the embedded ternary that only happens if there is only one item a little hard to follow.

Maybe something like:
```python
    # Create output mask filenames
    if num_imagesets == 1:
        filenames = [params.output.mask]
    else:
        # If there is more than one imageset, append a number to each output mask filename
        name, ext = os.path.splitext(params.output.mask)
        pad = len(str(num_imagesets))
        filenames = [
            "{name}_{num:0{pad}}{ext}".format(name=name, num=i + 1, pad=pad, ext=ext)
            for i in range(num_imagesets)
        ]
```
or, if you feel the format expansion doesn't need the naming:
```python
        filenames = [
            "{}_{:0{}}{}".format(name, i + 1, pad, ext) for i in range(num_imagesets)
        ]
```
There's possibly an argument that this could be moved off into a small internal function but I'm ambivalent about that in this case.

One other minor point: the `d` typing in format is redundant - if you are passing in a standard integer then it's the default, and it won't e.g. cause a stringified integer or float to convert. I guess there's a question of why it's in the standard if you never need it, but I'd imagine

1. printf-style has it (as printf *needs* the type in the string to correctly decode) and kept for sanity
2. You may want e.g. a custom class that accepts integer-style formatting but defaults to hex or something

As an aside, *knowing* about f-strings really makes me want to write them:
```python
filenames = [f"{name}_{i+1:0{pad}}{ext}" for i in range(num_imagesets)]
```